WIP - first attempt to store encrypted field values in a dedicated Co…

svendecabooter opened the pull request:
…ntentEntity

As mentioned on IRC, this is some WIP code that tries to implement the approach laid out in https://github.com/d8-contrib-modules/field_encrypt/issues/24.

Upon the first save of a node, the fields don't actually get encrypted and will return empty.
But when you update the node, things work roughly: the real field data is set to [ENCRYPTED] currently (still need to figure out why NULL or '' doesn't work) - upon loading the entity, this is dynamically replaced with the decrypted value from the EncryptedFieldValue entity.

The batch encrypting / decypting of values when field storage settings are changed, is also pretty broken due to these updates. To be fixed once the general approach works well.

svendecabooter committed:
WIP - first attempt to store encrypted field values in a dedicated ContentEntity

svendecabooter committed:
Make sure fields get properly encrypted on entity_insert + add hook_entity_delete support + code cleanup

svendecabooter commented:
Code is now fixed for inserts, so encryption / decryption of text field works.

Currently we still have to set a value for the storage of the unencrypted field (in current code this is set to "[ENCRYPTED]").
If we don't do this, the call to $field->getValue() in FieldEncryptProcessEntities processField() will return an empty array when decrypting, when it should be an array containing the properties.
i.e. we get:
`$field_value = [];`
instead of:
`$field_value = [
  'value' => '',
  'summary => '',
  'format => 'basic_html',
]`

We could try to recreate this array by getting a default value for the field or loading the correct properties, but there will still be a problem with properties that didn't get encrypted.

In the case above, when encrypting a text_with_summary field with format "full_html" and not setting any value for the unencrypted storage, we could retrieve the encypted text value, but the format setting would be gone, since no database entry was saved in the unencrypted field storage table.

So from what I can see so far, we'd have to set some placeholder value for things to keep working. Maybe an alternative approach could fix this, but I can't see how to do this right now.

Still TODO:
- Discuss saving of placeholder value
- Test with more field types
- Implement revision support
- Implement multilingual field support
- Fix batch un/re-encrypting of field values when field storage settings are changed
- Write unit tests
- Write functional tests

svendecabooter committed:
Add support for encrypting revisions

svendecabooter committed:
Add support for entities that do not allow revisions.

svendecabooter committed:
Add support for translatable entities.

svendecabooter commented:
The latest commits fix the following TODOs:
- Implement revision support
- Implement multilingual field support

(You'll have to reinstall the module for this to work)

I'd like to get some feedback before I tackle the following TODO:
- Fix batch un/re-encrypting of field values when field storage settings are changed

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
-    elseif ($form_encryption === 0) {
-      $field_encrypt_process_entities->decryptStoredField($field_entity_type, $field_name);
-    }
+    // @TODO: refactor logic for re-encrypting existing fields - currently broken, so commented out.
```
Thank you for leaving the comment here

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+function field_encrypt_entity_insert(Drupal\Core\Entity\EntityInterface $entity) {
+  // Save the entity again, but now encrypted.
+  // @TODO: check if there's not a cleaner way to do this.
+  $saved_entity = entity_load($entity->getEntityTypeId(), $entity->id());
```
Why do we have to do an entity load here? What context is missing from $entity?

svendecabooter replied:
The $entity variable passed as parameter does not have an ID assigned yet.
So $entity->id() returns no result, which prevents us from saving EncryptedFieldValue entities that reference that entity ID.

Perhaps we can get the entity ID in another way, but I haven't yet figured out how.

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
-   * @var $field_encrypt_process_entities \Drupal\field_encrypt\FieldEncryptProcessEntities
-   */
+function field_encrypt_entity_encrypt(Drupal\Core\Entity\EntityInterface $entity) {
+  if (($entity instanceof Drupal\field_encrypt\Entity\EncryptedFieldValueInterface) || !($entity instanceof Drupal\Core\Entity\ContentEntityInterface)) {
```
Can you leave some comments as to why this IF statement is here? It seems odd to me we need to check

svendecabooter replied:
It makes sure to only encrypt ContentEntities, not ConfigEntities.
Not sure if the code would work for ConfigEntities though, and if that would desired behaviour...

I also check if the entity being inserted / updated isn't a EncryptedFieldValue entity, since there seems little reason to encrypt the entity that stores the encrypted data. So it prevents the FieldEncryptProcessEntities service from being needlessly called and doing checks, if we already know it won't be necessary.

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+  // Check if this entity has fields that should be encrypted.
   $field_encrypt_process_entities = \Drupal::service('field_encrypt.process_entities');
-  $field_encrypt_process_entities->encryptEntity($entity);
+  if (!$field_encrypt_process_entities->checkEntity($entity)) {
```
I think this should be "checkEntityForEncryptedFields"

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+  }
+
+  // Set a flag to only perform the encryption once.
+  if ($entity->encrypting != TRUE) {
```
Are we processing fields? Shouldn't this be "processedFieldEncryption" where the status is set to TRUE after the processing is performed?

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+ * Encrypt entity fields before they are saved.
+ */
+function field_encrypt_entity_presave(\Drupal\Core\Entity\EntityInterface $entity) {
+  if (($entity instanceof Drupal\field_encrypt\Entity\EncryptedFieldValueInterface) || !($entity instanceof Drupal\Core\Entity\ContentEntityInterface)) {
```
This code is reused. 

I am guessing we move this into some checking function, like "checkEntityForFieldEncryption" or something ..... I'm not fully sure I understand what we're doing here

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
-    if (!($entity instanceof Drupal\Core\Entity\ContentEntityInterface)) {continue;}
-    $field_encrypt_process_entities->decryptEntity($entity);
+  foreach ($entities as &$entity) {
+    if (!($entity instanceof Drupal\Core\Entity\ContentEntityInterface)) {
```
I think this is yet another example of these checking function I mentioned above

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+      continue;
+    }
+    if (!$field_encrypt_process_entities->checkEntity($entity)) {
+      continue;
```
Stylistically, I'm not a huge fan of the "continue" syntax.

Ideally, we just wrap this in one checking function

nerdstein commented on field_encrypt.module:
```diff
@@ -131,44 +124,115 @@ function field_encrypt_form_field_add_form_builder($entity_type, \Drupal\field\E
+ * Remove EncryptedFieldValues associated with this entity.
+ */
+function field_encrypt_entity_delete(Drupal\Core\Entity\EntityInterface $entity) {
+  if (!($entity instanceof Drupal\Core\Entity\ContentEntityInterface)) {
```
See reusable check comments from before

nerdstein commented on field_encrypt.services.yml:
```diff
@@ -1,4 +1,8 @@
-    arguments: ['@entity.query', '@entity.manager', '@encryption', '@encrypt.encryption_profile.manager']
+    arguments: ['@entity.query', '@entity.manager', '@encryption', '@encrypt.encryption_profile.manager', '@field_encrypt.encrypted_field_value_manager']
+
+  field_encrypt.encrypted_field_value_manager:
```
Can you explain conceptually why this is a separate service from the one noted above? 

What is each one's domain?

nerdstein replied:
Disregard - I figured this out from the code below

nerdstein commented on src/Entity/EncryptedFieldValue.php:
```diff
@@ -0,0 +1,93 @@
+      ->setSetting('unsigned', TRUE);
+
+    $fields['field_name'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Field name'))
```
It may be excessive to store the field name, but I have no problem with it

svendecabooter replied:
It seems a requirement to me.
When you have an entity with 5 encrypted text fields, you wouldn't know which encrypted value is linked to what field on the entity, if we left out a reference to the fieldname.

nerdstein commented on src/Entity/EncryptedFieldValueInterface.php:
```diff
@@ -0,0 +1,34 @@
+   * @param string $value
+   *   The encrypted value.
+   */
+  public function setEncryptedValue($value);
```
I wonder if we should have a removeEncryptedValue() function if someone turns off Field Encrypt settings

nerdstein commented on src/FieldEncryptProcessEntities.php:
```diff
@@ -101,26 +135,72 @@ public function decryptEntity(ContentEntityInterface $entity) {
+      // Return value to store for unencrypted property.
+      // We can't set this to NULL, because then the field values are not saved,
+      // so we can replace them with their unencrypted value on load.
+      return '[ENCRYPTED]';
```
This threw me for a loop at first but I totally understand this based on your comments. Thank you for including it

nerdstein commented on src/FieldEncryptProcessEntities.php:
```diff
@@ -101,26 +135,72 @@ public function decryptEntity(ContentEntityInterface $entity) {
+   * @return bool
+   */
+  protected function checkField(FieldItemListInterface $field) {
+    if (!is_callable([$field, 'getFieldDefinition'])){
```
Wow, would be cool to have some comment here. I didn't realize this would even need to be done! :)

svendecabooter replied:
Perhaps these checks can be removed. They were in the original code I iterated on.

getFieldDefinition() is a method on ContentEntityBase, so I think it's fairly safe to assume the function will be callable.

We should verify that this doesn't result in errors in same edge cases, before removing though.

nerdstein commented on src/FieldEncryptProcessEntities.php:
```diff
@@ -202,24 +277,30 @@ protected function processEntity(ContentEntityInterface $entity, $op = 'encrypt'
-    foreach ($entity->getFields() as $field){
-      $this->processField($field, $op);
+    // Process all language variants of the entity.
+    $languages = $entity->getTranslationLanguages();
```
Awesome

nerdstein commented on src/FieldEncryptProcessEntitiesInterface.php:
```diff
@@ -30,6 +30,16 @@ public function encryptEntity(ContentEntityInterface $entity);
+   * @return bool
+   *   TRUE if entity has encrypted fields, FALSE if not.
+   */
+  public function checkEntity(ContentEntityInterface $entity);
```
Please rename from comments left above

nerdstein commented:
I have finished reviewing it thus far and have left comments. Overall, it's great. 

Please review my comments. I can merge whenever you would like (if you think we need to get the code in and iterate), or I can merge after you update from my comments.

I will say there is a lot of code commented out. I would really like to see "refactor" lines done before we merge, if possible.

svendecabooter committed:
clean up code

svendecabooter commented:
Functionality added in the latest commit:
- remove double interface checks and continue statements - replace with reusable function field_encrypt_allow_encryption().
- changed the "encrypting" flag to "doFieldEncryption", as it better represents what it's for.
- renamed checkEntity() to entityHasEncryptedFields()
- general code cleanup according to Drupal coding standards
- reorder methods in FieldEncryptProcessEntities to follow the logical flow

Better suggestions for naming then the ones above are always welcome :)

nerdstein commented:
Looks good. I'm merging and hoping we can continue with the remaining effort.

I think we very quickly need to get into the caching systems
